java/arcs/android/storage/service/BindingContext.kt

```diff
@@ -190,7 +217,47 @@ class BindingContext(
         }
     }
 
+    /**
+     * Increment an active message count and send it to the active messages count flow.
+     *
+     * Calls to [idle] will wait for this value to reach 0 before checking store activity.
+     *
+     * Avoid using this for calls to [idle] itself, so as to not increment the activity counter.
+     */
+    private fun launchNonIdleAction(action: suspend () -> Unit) {
+        activeMessagesCountFlow.increment()
+        // TODO(b/168724138) - Switch to a channel/flow queue-based approach.
+        actionQueue
+            .launch(action)
+            .invokeOnCompletion { activeMessagesCountFlow.decrement() }
+    }
+
     companion object {
         private val nextId = atomic(0)
     }
 }
+
+@ExperimentalCoroutinesApi
+/**
+ * A simple helper class to provide serialized operation execution on the provided
+ * [CoroutineScope].
+ *
+ * TODO(b/168724138) - Switch to a channel/flow queue-based approach and remove this.
+ */
+class MutexOperationQueue(
+    private val scope: CoroutineScope
+) {
+    // A mutex to serialize the incoming actions.
+    // Since coroutine mutexes process lock-waiters in order, this services as a quick-and-dirty
+    // queue for us.
+    private val actionMutex = Mutex()
+
+    fun launch(action: suspend () -> Unit): Job {
```

Comment:
Looks like the `launch` is serializing storage service calls per stores (storage keys).
Note: could impact performance when a handle (storage keys) is used in multiple Arcs.

Reply:
I think we can address this by introducing a new serialization queue for each channel ID encountered.

Reply:
I'm probably wrong.
The same handle (storage key) in two Arcs -> two EntityHandleManager -> two StorageProxy under corresponding EntityHandleManager.
Each StorageProxy binds to the service so two BindingContext instances.
So performance is not a problem as the flow/queue is tied to each BindingContext instance.

Reply:
There is only one instance of the BindingContext per storage key, though.

Reply:
!! How many `onBind` got called for two StorageProxy at two Arcs?

Reply:
Hidden in a random paragraph here: https://developer.android.com/guide/components/bound-services#bind-started-service
```
You can connect multiple clients to a service simultaneously. However, the system caches the IBinder service communication channel. In other words, the system calls the service's onBind() method to generate the IBinder only when the first client binds. The system then delivers that same IBinder to all additional clients that bind to that same service, without calling onBind() again.
```

This is why we include the `StorageKey` as part of the `ACTION` in the bind Intent; Android will use `ACTION` as part of the key for the binder cache, but it doesn't use any "extras". 

It led to some confusing behavior, and I find it's not really very well represented in the docs. 

Another confusing in the docs for `onBind`:
`Note that any extras that were included with the Intent at that point will not be seen here.`

That was clearly not *exactly* true, since we pass options as an extra, and we definitely see them... But what it does mean is that for any given storage key, only the extras provided the first time we bind to the service will be seen.

Reply:
Right, I knew AiAiStorageService overrides the intent creation and thought it was intentionally causing onBind gets called for every service binding. Checking it further, yes it only used storage key as `action`.
ServiceManager manages the binder cache in a hashmap (perhaps not up to date but at least what I knew).
And the map key is exactly `action`.